### PR TITLE
Integrate C++ ChunkDataset with Pytorch (Python) Dataloader

### PR DESCRIPTION
**Pull request by thiagocrepaldi**

### CODE REVIEW

**Review by ssnl — COMMENTED** [review]
Would it be possible to do this after #19228 is merged? That would make things much less hacky.

**Comment by soumith** [comment]
yes i think we should base this on top of #19228

**Comment by thiagocrepaldi** [comment]
Sure, this is just a draft version yet - dint know Drafts were tracked at all :)

File: torch/utils/data/dataloader.py
Comment (SsnL) [inline torch/utils/data/dataloader.py]: `auto_collation` is on if either `batch_size` or `batch_sampler` is specified.

Also, why is this needed?

Comment (thiagocrepaldi) [reply torch/utils/data/dataloader.py]: We are creating python bindings for the c++ chunkdataset API implementation already on pytorch code base (chunk.h). On that dataset implementation, a full batch is returned from the dataset (regardless of batch_sampler, sampler, batch_size, etc of python DataLoader class). 

Another way to go would be set batch_size=None on the Dataloader ctor, but that would be misleading for the users, as ChunkDataset receives batch_size as input and it is a positive number

Comment (thiagocrepaldi) [reply torch/utils/data/dataloader.py]: I will provide a test soon and this use will be clearer, but any feedback is welcome

**Comment by thiagocrepaldi** [comment]
At https://discuss.pytorch.org/t/chunked-data-randomization/45162, there is a user interested in using ChunkDataset python bindings

File: torch/csrc/api/include/torch/data/init.h
Comment (xzhu1900) [inline torch/csrc/api/include/torch/data/init.h]: Maybe we can consider adding some comment explaining what current_stride_ and total_stride_ do? It might be confusing for people first look at it.

File: torch/csrc/api/src/data/init.cpp
Comment (xzhu1900) [inline torch/csrc/api/src/data/init.cpp]: Shall we move init_dataset_binding_ to test directory? I think we should avoid init test only binding in product code.

File: test/test_dataloader.py
Comment (thiagocrepaldi) [reply test/test_dataloader.py]: Linter formatted the file according to pep8 standard

File: torch/csrc/api/include/torch/data/init.h
Comment (thiagocrepaldi) [reply torch/csrc/api/include/torch/data/init.h]: Great idea. Adding documentation to SamplerWrapper class, which includes strides

File: torch/csrc/api/src/data/init.cpp
Comment (thiagocrepaldi) [reply torch/csrc/api/src/data/init.cpp]: `init_dataset_bindings_test` doesnt really implement any test - bad naming alert. It creates a python binding that can be used  as an example for future implementations. Adding to this document-by-example scenario approach, the bindings are also used used by a Python tests (test_dataloader.py) to illustrate how a Python script would use the new Python API based on bindings for the existing C++ code. The actual python test is just a tiny bonus that came with the package :)

Maybe renaming `init_dataset_bindings_test` to `init_dataset_bindings_example` makes this separation between core code and example clearer. What do you think? I will add comments on the new revision

It is not possible to do bindings on main torch module from test folder without

File: test/test_dataloader.py
Comment (jaliyae) [inline test/test_dataloader.py]: You only need these members to create the wrappers, so can we simply create them in the constructor of the wrappers.

File: torch/csrc/api/include/torch/data/init.h
Comment (jaliyae) [inline torch/csrc/api/include/torch/data/init.h]: Please change this to set_total_strides() instead of stride.

Also, regarding the warning. I would explain, the total strides and the current stride depend on the cluster setup and the number of workers used in each dataloader, so the reset will not change them.

File: torch/csrc/api/include/torch/data/init.h
Comment (jaliyae) [inline torch/csrc/api/include/torch/data/init.h]: total_strides()

File: test/test_dataloader.py
Comment (thiagocrepaldi) [reply test/test_dataloader.py]: Which one?

`chunk_count` is used by `chunk_sampler`
`batch_size` is used by `example_sampler` and `ChunkDatasetOptions`
`cache_size` and `preloaders` are used by ChunkDatasetOptions

`SamplerWrapper` only gets the samplers instances (for chunk_sampler and example_sampler), no other input is used by the wrapper

Both chunk and example samplers can be created with different parameters depending on which Sampler specialization they use (sequential, random, distributed sequential, distributed random, etc), so it is not possible to embed all parameters in the wrapper class

File: torch/csrc/api/include/torch/data/init.h
Comment (thiagocrepaldi) [reply torch/csrc/api/include/torch/data/init.h]: good idea.

File: torch/csrc/api/include/torch/data/init.h
Comment (thiagocrepaldi) [reply torch/csrc/api/include/torch/data/init.h]: good idea

File: torch/csrc/api/include/torch/data/init.h
Comment (jaliyae) [inline torch/csrc/api/include/torch/data/init.h]: all even

**Review by jaliyae — COMMENTED** [review]
Please fix the comments and let's make this a PR.

File: torch/utils/data/_utils/worker.py
Comment (SsnL) [inline torch/utils/data/_utils/worker.py]: This can absolutely not happen. You need BC-preserving mechanism here. Also, why do you even need this? You can get the dataset object from `torch.utils.data.get_worker_info`

File: torch/utils/data/dataloader.py
Comment (SsnL) [inline torch/utils/data/dataloader.py]: Why is this needed? Can't you just use a `collate_fn` with these args? E.g., `functools.partial`?

File: torch/utils/data/dataloader.py
Comment (SsnL) [reply torch/utils/data/dataloader.py]: Thanks for the explanation. But I am still not convinced that adding this is better. It complicates the already pretty complicated auto-collation behavior. I.e., previously, auto collating means that the dataset is yielding individual samples, but now it doesn't!

Actually, for ChunkDataset, I don't think that you should use any of the DataLoader's auto-collation mechanism, including `collate_fn`. It should be clear that the dataset yields batched samples, and thus the dataloader should not do any collation. If you put `batch_size` in one place (dataset) and `collate_fn`+`auto_collation` in another (dataloader), you are essentially splitting the batching logic into two different places. It is against the philosophy of current dataloader design that the dataloader either handles all collation or not, so I think it would just be more confusing.

My suggestion is to just delegate all batching/collation to the `ChunkDataset`, i.e., putting `batch_size` and `collate_fn` to its ctor. And when one uses a `ChunkDataset`, they should understand what it does and set in dataloader ctor `batch_size=None`.

**Review by soumith — CHANGES_REQUESTED** [review]
@ssnl has some comments that are blocking, with requesting changes.
Once they have been resolved and addressed, dismiss my review so that it can get back into review queue

File: torch/utils/data/_utils/worker.py
Comment (thiagocrepaldi) [reply torch/utils/data/_utils/worker.py]: In the context of ChunkDataset on multiple workers, each worker needs to have access to the dataset, so it can configure its internal data reading engine. Without that, each worker would read the same data.

`torch.utils.data.get_worker_info` is a good idea! The dataset argument was presenting the original PR, before the new DataLoader implementation - I forgot to use it. Thanks for pointing it out!

File: torch/utils/data/dataloader.py
Comment (thiagocrepaldi) [reply torch/utils/data/dataloader.py]: `collate_fn` is being used to transform batches before converting them to Tensors and return. Some transformations require parameters.

I considered using `functors.partial` to wrap arguments on collate_fn, but I wasn't sure whether that would be the simpler solution for the users.

If you think so, I can do this change

File: torch/utils/data/dataloader.py
Comment (thiagocrepaldi) [reply torch/utils/data/dataloader.py]: Just as a context, this PR is roughly divided in two parts: 

- Adds new python bindings for existing C++ API (Chunkdataset was merged since Pytorch 1.1)
- Enhances Python Dataloader implementation to support the new ChunkDataset

This approach is interesting because C++ ChunkDataset API can support efficient data reading while python Dataloader allows an already familiar API for pytorch developers and flexible data transformation (such as torchvision.Transforms, numpy, etc) through `collate_fn`. Because some data transformation are very complicated to be done on C++, python bindings and integration with its DataLoader is a natural step for ChunkDataset. 

As you may have noticed, enhancing `auto-collation` is at the heart of the second part of the PR. I do agree DataLoader is already complex and, with that in mind, if `auto_collation` is not explicitly overridden by the user, the default behavior is exactly the same as before. That was done to prevent `DataLoader` from getting any more complicated for most users, but also to allow a finer granularity for advanced scenarios such as the new ChunkDataset API.

As you suggested, I am trying to disable auto-collation for `ChunkDataset(IterableDataset)`. By default, auto-collation is always on for `IterableDataset` as `batch_size=1` as per `DataLoader.__init__` (dataloader.py:116). As a consequence, `batch_sampler` will be assigned to `BatchSampler(sampler, batch_size, drop_last)` (dataloader.py:211). Because of that, `_auto_collation`property will return `True` (dataloader.py:243) and auto-collation will take place for `_IterableDatasetFetcher.fetch` (fetch.py:23). In the current implementation, the only way to prevent this from happening is to assign `batch_size=None`on DataLoader ctor, but that can be confusing for users of ChunkDataset API, as it does require a valid batch_size on its ctor. Given all the data flow for the current DataLoader implementation, my approach was to allow a manual toggle that disables auto-collation, preventing a confusing batch_size=None for Dataloader and batch_size>0 for ChunkDataset in the same script.

Regarding `collate_fn`, I have replied to it in another comment below (at dataloader.py:88)

@SsnL  Given this very long text explaining the intended data flow for `ChunkDataset` (sorry about that), do you still think that using two different values for `batch_size` (one `DataLoader ctor` and another for `ChunkDataset ctor`) is better than providing a setter for `Dataloader.auto_collation` that allows disabling it while keeping the default behavior intact?

**Comment by thiagocrepaldi** [comment]
> @SsnL has some comments that are blocking, with requesting changes.
> Once they have been resolved and addressed, dismiss my review so that it can get back into review queue

Thank you, @soumith. I will work closely with @SsnL and we hopefully can have this working before 1.2 release :)

File: torch/utils/data/dataloader.py
Comment (apaszke) [reply torch/utils/data/dataloader.py]: Yes, we should not be in the business of managing that. If someone needs extra parameters, they can either make a closure, or use `functools.partial`.

File: torch/utils/data/dataloader.py
Comment (apaszke) [inline torch/utils/data/dataloader.py]: Can you elaborate on what this does? I'm not sure if we really need this argument. We now support batch datasets on the Python side.

**Review by apaszke — CHANGES_REQUESTED** [review]

File: torch/utils/data/dataloader.py
Comment (fmassa) [reply torch/utils/data/dataloader.py]: Or [use a class with a `__call__` method](https://github.com/facebookresearch/maskrcnn-benchmark/blob/55796a04ea770029a80cf5933cc5c3f3f6fa59cf/maskrcnn_benchmark/data/collate_batch.py#L5-L20)

Comment (thiagocrepaldi) [reply torch/utils/data/dataloader.py]: Agreed.

File: torch/utils/data/_utils/worker.py
Comment (thiagocrepaldi) [reply torch/utils/data/_utils/worker.py]: Agreed. Using `torch.utils.data.get_worker_info().dataset`

File: torch/utils/data/dataloader.py
Comment (thiagocrepaldi) [reply torch/utils/data/dataloader.py]: With current `DataLoader ctor` default values, `IterableDataset` dataset will always be on auto-collation mode. Because it is possible that some dataset implementations, especially the new `ChunkDataset` which extends `IterableDataset` not necessarily need auto-collation, I have added this setter (before there was a getter only) to override this behavior. Another approach would be to explicitly set `batch_size=None`.

I preferred the former approach because `ChunkDataset` ctor requires positive `batch_size` and setting `batch_size=None` to DataLoader class could be confusing for the user. In the same script we would have batch_size with a positive value and None

File: torch/utils/data/dataloader.py
Comment (thiagocrepaldi) [reply torch/utils/data/dataloader.py]: As per your suggestion, I have added a `collate_fn` to `ChunkDataset` class, which is called upon its `__next__` method.

Comment (thiagocrepaldi) [reply torch/utils/data/dataloader.py]: Removed auto_collation argument too

File: torch/utils/data/dataloader.py
Comment (thiagocrepaldi) [reply torch/utils/data/dataloader.py]: Removed auto_collation in the latest revision

**Comment by thiagocrepaldi** [comment]
All suggestions made by @apaszke and @SsnL are now addressed as requested

File: torch/utils/data/dataset.py
Comment (apaszke) [inline torch/utils/data/dataset.py]: Why is it the dataset that's handling the collation? This should probably be a functionality separated from this.

File: torch/utils/data/dataset.py
Comment (apaszke) [inline torch/utils/data/dataset.py]: This is an implementation detail. Is it limited to that kind of dataset only? If yes, why aren't we checking that the user hasn't provided us with an object we don't expect to see?

File: torch/utils/data/dataset.py
Comment (apaszke) [reply torch/utils/data/dataset.py]: How does one even create a chunk dataset in Python? We cannot expect people to look into the `torch._C` namespace --- it's private!

File: torch/csrc/api/src/data/init.cpp
Comment (apaszke) [inline torch/csrc/api/src/data/init.cpp]: I'd really like us to avoid having to bundle this with our library and loading it by default since this is test-only code. Things like that make our binary unnecessarily large. Can we make this an extension or at least enable conditional loading?

File: torch/csrc/api/src/data/init.cpp
Comment (apaszke) [inline torch/csrc/api/src/data/init.cpp]: There is a `DistributedSampler` in Python, so why are we binding it in here??

File: torch/csrc/api/src/data/init.cpp
Comment (apaszke) [inline torch/csrc/api/src/data/init.cpp]: Same for those. They're already available in Python

**Review by apaszke — COMMENTED** [review]
Can you please provide some additonal description of this PR? I'm not sure which part is supposed to be implemented in Python and which in C++? Do you expect `ChunkReader`s to be in Python? Do you expect full interoperability with Python samplers? How is a user supposed to create a `ChunkReader` on their own? There's hardly any documentation and the test makes use of the private `torch._C` namespace to get anything done. This definitely needs to get polished before we can land it.

File: torch/utils/data/dataset.py
Comment (thiagocrepaldi) [reply torch/utils/data/dataset.py]: Regarding the first comment, I agree that using existing `DataLoader` (dataloader.py) `collate_fn` is a cleaner way. The new revision already cotnains that change

Comment (thiagocrepaldi) [reply torch/utils/data/dataset.py]: Regarding the second comment, I have created a public namespace for the chunk dataset. Now the user can `import torch.utils.data.chunk` and get the same functionality

I also created a chunk.rst page which describes the API and give two examples on how to use it. The first one is the simplest way and the second the most advanced scenario.

File: torch/utils/data/dataset.py
Comment (thiagocrepaldi) [reply torch/utils/data/dataset.py]: Yes, ChunkDataset is implemented based on `IterableDataset` because the size of the dataset is not known

We cant do any type checking because the C++ ChunkDataset API is based on templating (as much of Pytorch C++), which makes impossible to check if a specific ChunkDataset (e.g. DummyChunkDataset) is an instance of generic (templatized) `ChunkDataset ` class

File: torch/csrc/api/src/data/init.cpp
Comment (thiagocrepaldi) [reply torch/csrc/api/src/data/init.cpp]: Great idea. I have moved this code to the existing cpp extension test and also added unit tests there too!

File: torch/csrc/api/src/data/init.cpp
Comment (thiagocrepaldi) [reply torch/csrc/api/src/data/init.cpp]: ChunkDataset is a C++ API (exposed to python) and it only accepts C++ samplers (exposed to python). We can't use a Python sampler on a C++ Dataset

File: torch/csrc/api/src/data/init.cpp
Comment (thiagocrepaldi) [reply torch/csrc/api/src/data/init.cpp]: Same as before. We need a c++ sampler for a c++ dataset object

**Comment by thiagocrepaldi** [comment]
Gentle ping :)

**Review by apaszke — CHANGES_REQUESTED** [review]
Sorry for the delay @thiagocrepaldi. I'm afraid this is still not ready to get in. The docs are severely missing. They're really meant to be a textual description of how the system is expected to work from the user side, but right now they're mostly a list of samplers and two huge code snippets with no text in between.

More importantly, I'm very wary of adding this to the core. From what I understand the C++ utilities you expose are completely non-composable with the Python implementations, right? Like, we end up having two disjoint sets of samplers that do the same thing, but don't really function in the same way and are expected to be passed in to different projects. This is very much against the principles we've been following to guide the development of our APIs, because it makes it very hard for users to understand when should they use each of them. I think that unless we have a good integration and deduplication story (and unless it is documented properly) we should not move forward with this patch. If you want you can always provide this functionality as a separate C++ extension that others can install, because all the necessary headers should be shipped along with the library.

I know this probably sounds quite disappointing, but I believe this is the best solution long-term. I'm happy to discuss further if you disagree.

**Comment by thiagocrepaldi** [comment]
> Sorry for the delay @thiagocrepaldi. I'm afraid this is still not ready to get in. The docs are severely missing. They're really meant to be a textual description of how the system is expected to work from the user side, but right now they're mostly a list of samplers and two huge code snippets with no text in between.
> 
> More importantly, I'm very wary of adding this to the core. From what I understand the C++ utilities you expose are completely non-composable with the Python implementations, right? Like, we end up having two disjoint sets of samplers that do the same thing, but don't really function in the same way and are expected to be passed in to different projects. This is very much against the principles we've been following to guide the development of our APIs, because it makes it very hard for users to understand when should they use each of them. I think that unless we have a good integration and deduplication story (and unless it is documented properly) we should not move forward with this patch. If you want you can always provide this functionality as a separate C++ extension that others can install, because all the necessary headers should be shipped along with the library.
> 
> I know this probably sounds quite disappointing, but I believe this is the best solution long-term. I'm happy to discuss further if you disagree.

Thanks, for the feedback, @apaszke !

Ok, I agree that the two types of Samplers can be confusing. So I've got rid of them!
I also added two helper functions `bind_chunkdataset` and `bind_chunkdatareader` that do most of the heavy lifting to expose new readers. Finally, I've improved the documents the reflect the changes and add more textual content. Let me know what do you think!

File: docs/source/chunk.rst
Comment (apaszke) [inline docs/source/chunk.rst]: This is a highly non-Pythonic interface. If you want to integrate with the core, it's not enough to simply mimic the awkward code patterns that C++ sometimes enforces on us. We don't use `*Options` classes for anything. Those arguments should be inlined directly in the constructor.

File: docs/source/chunk.rst
Comment (apaszke) [inline docs/source/chunk.rst]: Why do you have to wrap a `ChunkDataset` that's been exposed to Python into another object? This is again a sign of a surprising API. All the necessary methods should be implemented in the pybind's base class such that subclasses automatically can work with Python data loaders.

File: docs/source/chunk.rst
Comment (apaszke) [inline docs/source/chunk.rst]: Did you really want to spawn a single new process? If you don't want to create any workers just don't specify the argument

File: docs/source/chunk.rst
Comment (apaszke) [inline docs/source/chunk.rst]: Didn't the previous example use the extension API already?

File: docs/source/chunk.rst
Comment (apaszke) [inline docs/source/chunk.rst]: This has already been mentioned. I don't think you need to repeat this.

File: docs/source/chunk.rst
Comment (apaszke) [inline docs/source/chunk.rst]: I really think that given the ratio of code to text in here, this should really become an example repository and not part of our docs...

File: docs/source/chunk.rst
Comment (apaszke) [reply docs/source/chunk.rst]: Plus, if I understand correctly this is called `ChunkDatasetWrapper`.

File: test/test_cpp_extensions.py
Comment (apaszke) [inline test/test_cpp_extensions.py]: Does this really have to be this complex? Is that the minimal amount of code you need to test something? What are you trying to check in here?

File: torch/csrc/api/include/torch/data/init.h
Comment (apaszke) [inline torch/csrc/api/include/torch/data/init.h]: This is an implementation detail that's not even exposed anymore

File: torch/csrc/api/include/torch/data/init.h
Comment (apaszke) [inline torch/csrc/api/include/torch/data/init.h]: I just wanted to double check that you're aware that this means that you will never be able to use other samplers with this kind of dataset from Python without breaking backwards compatibility?

File: docs/source/chunk.rst
Comment (apaszke) [inline docs/source/chunk.rst]: Copy paste?

**Review by apaszke — COMMENTED** [review]
The tutorial contains many errors. As I've said it should be a repository so that you can verify that the code runs and doesn't have typos.

In any case, without the duplicated sampler hierarchy this is more reasonable, but I'm still wondering if this thing wouldn't be better off as a separate library... It could expose `bind_chunkdatareader` and `bind_chunkdataset`, and could be used by extensions to expose their own datasets.

Like, the interface proposed in this PR is quite constrained, and requires fiddling with custom C++ extensions to expose custom datasets anyway, so why can't you just make it easier for people to write those extensions without putting things into the core library? I just don't see the urgency of having this in the core.

**Comment by thiagocrepaldi** [comment]
Thanks for the feedback, @apaszke . based on your comments, I will close this PR, discuss it internally and reopen (if and when appropriate)

**Comment by apaszke** [comment]
Thanks! I'm sorry about that, but I really don't feel like this is an addition that absolutely needs to be in the core. We're trying to keep it lean because every line and feature increases the maintenance burden on us and slows us down. Hope you understand.

**Pull request closed without merging**
